src/hotspot/cpu/x86/gc/g1/g1_x86_64.ad

```diff
@@ -0,0 +1,370 @@
+                         $tmp2$$Register /* pre_val */,
+                         $tmp3$$Register /* tmp */,
+                         RegSet::of($tmp1$$Register, $src$$Register) /* preserve */);
+    __ movl(Address($tmp1$$Register, 0), $src$$Register);
+    if ((barrier_data() & G1C2BarrierPost) != 0) {
+      __ movl($tmp2$$Register, $src$$Register);
+      if ((barrier_data() & G1C2BarrierPostNotNull) == 0) {
+        __ decode_heap_oop($tmp2$$Register);
+      } else {
+        __ decode_heap_oop_not_null($tmp2$$Register);
+      }
+    }
+    g1_post_write_barrier(masm, this,
+                          $tmp1$$Register /* store_addr */,
+                          $tmp2$$Register /* new_val */,
+                          $tmp3$$Register /* tmp1 */,
+                          $tmp2$$Register /* tmp2 */);
+  %}
+  ins_pipe(ialu_mem_reg);
+%}
+
+instruct g1EncodePAndStoreN(memory mem, any_RegP src, rRegP tmp1, rRegP tmp2, rRegP tmp3, rFlagsReg cr)
+%{
+  predicate(UseG1GC && n->as_Store()->barrier_data() != 0);
+  match(Set mem (StoreN mem (EncodeP src)));
+  effect(TEMP tmp1, TEMP tmp2, TEMP tmp3, KILL cr);
+  ins_cost(125); // XXX
+  format %{ "encode_heap_oop $src\n\t"
+            "movl   $mem, $src\t# ptr" %}
+  ins_encode %{
+    __ lea($tmp1$$Register, $mem$$Address);
+    g1_pre_write_barrier(masm, this,
+                         $tmp1$$Register /* obj */,
+                         $tmp2$$Register /* pre_val */,
+                         $tmp3$$Register /* tmp */,
+                         RegSet::of($tmp1$$Register, $src$$Register) /* preserve */);
+    __ movq($tmp2$$Register, $src$$Register);
+    if ((barrier_data() & G1C2BarrierPostNotNull) == 0) {
+      __ encode_heap_oop($tmp2$$Register);
+    } else {
+      __ encode_heap_oop_not_null($tmp2$$Register);
+    }
+    __ movl(Address($tmp1$$Register, 0), $tmp2$$Register);
+    g1_post_write_barrier(masm, this,
+                          $tmp1$$Register /* store_addr */,
+                          $src$$Register /* new_val */,
+                          $tmp3$$Register /* tmp1 */,
+                          $tmp2$$Register /* tmp2 */);
+  %}
+  ins_pipe(ialu_mem_reg);
+%}
+
+instruct g1CompareAndExchangeP(indirect mem, rRegP newval, rRegP tmp1, rRegP tmp2, rRegP tmp3, rax_RegP oldval, rFlagsReg cr)
+%{
+  predicate(UseG1GC && n->as_LoadStore()->barrier_data() != 0);
+  match(Set oldval (CompareAndExchangeP mem (Binary oldval newval)));
+  effect(TEMP tmp1, TEMP tmp2, TEMP tmp3, KILL cr);
+  format %{ "lock\n\t"
+            "cmpxchgq $newval, $mem" %}
+  ins_encode %{
+    assert_different_registers($oldval$$Register, $mem$$Register);
+    g1_pre_write_barrier(masm, this,
+                         $mem$$Register /* obj */,
+                         $tmp2$$Register /* pre_val */,
+                         $tmp3$$Register /* tmp */,
+                         RegSet::of($mem$$Register, $newval$$Register, $oldval$$Register) /* preserve */);
```

Comment:
The only value which can get overwritten is `oldval`. Optimization idea: Pass `oldval` to the SATB barrier. There is no load of the old value required.

Reply:
Thanks, I will test and apply this one: it is simple enough and, in a way, even more intuitive than the current solution. For reference, I prototyped it [here](https://github.com/robcasloz/jdk/blob/JDK-8334060-g1-late-barrier-expansion-x64-optimizations/src/hotspot/cpu/x86/gc/g1/g1_x86_64.ad) (guarded by `UseOldValInPreBarriers`). It should also apply to `g1CompareAndSwapP`, right?

Reply:
Exactly.

Reply:
Done (commit 9211280).